### PR TITLE
JDK-8310066: Improve test coverage for JVMTI GetThreadState on carrier and mounted vthread

### PR DESCRIPTION
**Pull request by alexmenkov**
This is follow-up JDK-8307153/JDK-8309612 (JVMTI GetThreadState on carrier should return STATE_WAITING)
New test tests GetThreadState for different thread states.
The test detected a bug in the implementation, new issue is created: JDK-8310584
Corresponding testcases are commented now in the test, fix for JDK-8310584 will uncomment them

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310066](https://bugs.openjdk.org/browse/JDK-8310066): Improve test coverage for JVMTI GetThreadState on carrier and mounted vthread (**Enhancement** - P4)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [8918411b](https://git.openjdk.org/jdk/pull/14622/files/8918411bdaa4882ba998a9ba4ae777273a0e54b2)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14622/head:pull/14622` \
`$ git checkout pull/14622`

Update a local copy of the PR: \
`$ git checkout pull/14622` \
`$ git pull https://git.openjdk.org/jdk.git pull/14622/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14622`

View PR using the GUI difftool: \
`$ git pr show -t 14622`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14622.diff">https://git.openjdk.org/jdk/pull/14622.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14622#issuecomment-1603590882)

### CODE REVIEW

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (turbanoff) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:85]:
```suggestion
        synchronized (syncObj) {
```

Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java]: Fixed

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:241]: Q: Where is this native method invoked?

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:117]: Shout this code account for spurious wakeups?

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp:176]: Is this loop reliable in terms of thread states?

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp:73]: Nit: lines 65, 73 have wrong indent.

**Review by sspitsyn — COMMENTED** [review]
The test is great.
I realize it is not easy to make it fully reliable though.
So, will need another pass.

**Comment by alexmenkov** [comment]
> The test is great. I realize it is not easy to make it fully reliable though. So, will need another pass.

Good point about spurious wakeups.
Updated the test to handle them.
Now for WAITING scenarios the test suspends test thread and ensures it's waiting. While the thread is suspended (either carrier or virtual) we can safely verify thread states

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:298]: Good catch. Test thread inNative testcase never exits, it's terminated on the test exit.
Fixed.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:130]: Fixed

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp:206]: The code is pure native (sleep_ms uses native functions, it does not interact with JVM), so there is no thread state changes

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp]: Fixed.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:41]: Nit: `each scenarios` => `each scenario`

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:62]: Q: It is not clear why an array is needed instead of non-final local. The same question for `waiting()` method below,

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java]: Fixed.

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:62]: It's used in lambda, so must be final
I think both "flag to stop" and "flag to continue" are clear, I prefer "flag to stop" as it does not require additional initialization

Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:62]: Okay, I see why an array is needed.
It is really minor and up to you. Just wanted to get rid of one negation. :)

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp
Comment (sspitsyn) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp:112]: Nit: I guess, error message at line 110 has to be: 
`ERROR: not all expected 'weak' bits are set`
Also, it looks like the check 3a is not really needed as the 3b should cover it.
But I leave it up to you, if you think check 3a gives some convenience.

**Review by sspitsyn — APPROVED** [review]
Thank you for the test. It is nice to have it.
It looks good.
Thanks,
Serguei

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp:112]: 'weak' value is for 'sleeping' testcase. In the case we have 2 valid states: JVMTI_THREAD_STATE_SLEEPING and JVMTI_THREAD_STATE_PARKED, but only one of them should be reported.
Test passes expected_weak == JVMTI_THREAD_STATE_SLEEPING | JVMTI_THREAD_STATE_PARKED.
check 3a verifies that at least 1 bit is set (JVMTI_THREAD_STATE_SLEEPING, JVMTI_THREAD_STATE_PARKED or both).
check 3b verifies that only one of the bits is set (i.e. we don't have both JVMTI_THREAD_STATE_SLEEPING and JVMTI_THREAD_STATE_PARKED)

**Comment by alexmenkov** [comment]
Ping. need one more reviewer

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:41]:
```suggestion
 * For each scenario the test also checks states after carrier and virtual threads suspend/resume
```

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java:45]:
```suggestion
 * To avoid this, the test thread should be suspended (i.e. carrier and/or mounted virtual thread is suspended).
```

**Review by plummercj — APPROVED** [review]
Just a couple of comment suggestions. Otherwise looks good.

What testing did you do? Any -Xcomp testing?

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java]: Fixed

File: test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java
Comment (alexmenkov) [reply test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java]: Fixed.

**Comment by alexmenkov** [comment]
> What testing did you do? Any -Xcomp testing?

Just run with -Xcomp on all mach5 platforms, passed

**Review by plummercj — APPROVED** [review]

**Pull request closed without merging**
